src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/EnumConverterTests.cs

```diff
@@ -554,5 +555,85 @@ public enum SampleEnumSByte : sbyte
+            JsonSerializerOptions options = new()
+            {
+                Converters =
+                {
+                    new JsonStringEnumConverter(namingPolicy: new SimpleSnakeCasePolicy() )
+                }
+            };
+
+            BindingFlags bindingFlags = JsonSerializer.Deserialize<BindingFlags>(@"""non_public""", options);
+            Assert.Equal(BindingFlags.NonPublic, bindingFlags);
+
+            // Flags supported without naming policy.
+            bindingFlags = JsonSerializer.Deserialize<BindingFlags>(@"""NonPublic, Public""", options);
+            Assert.Equal(BindingFlags.NonPublic | BindingFlags.Public, bindingFlags);
+
+            // Flags supported with naming policy.
+            bindingFlags = JsonSerializer.Deserialize<BindingFlags>(@"""static, public""", options);
+            Assert.Equal(BindingFlags.Static | BindingFlags.Public, bindingFlags);
+
+            // Null not supported.
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<BindingFlags>("null", options));
+
+            // Null supported for nullable enum.
+            Assert.Null(JsonSerializer.Deserialize<BindingFlags?>("null", options));
+        }
+
+        [Fact]
+        public static void EnumDictionaryKeyDeserialization()
+        {
+            JsonNamingPolicy snakeCasePolicy = new SimpleSnakeCasePolicy();
+            JsonSerializerOptions options = new()
+            {
+                Converters =
+                {
+                    new JsonStringEnumConverter(namingPolicy: snakeCasePolicy)
+                },
+                DictionaryKeyPolicy = snakeCasePolicy
+            };
+
+            // Baseline.
+            var dict = JsonSerializer.Deserialize<Dictionary<BindingFlags, int>>(@"{""NonPublic, Public"": 1}", options);
+            Assert.Equal(1, dict[BindingFlags.NonPublic | BindingFlags.Public]);
+
+            // DictionaryKeyPolicy not honored for dict key deserialization.
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Dictionary<BindingFlags, int>>(@"{""NonPublic0, Public0"": 1}", options));
+
+            // EnumConverter naming policy not honored.
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Dictionary<BindingFlags, int>>(@"{""non_public, static"": 0, ""NonPublic, Public"": 1}", options));
+        }
+
+        [Fact]
+        public static void EnumDictionaryKeySerialization()
+        {
+            JsonSerializerOptions options = new()
+            {
+                DictionaryKeyPolicy = new SimpleSnakeCasePolicy()
+            };
+
+            Dictionary<BindingFlags, int> dict = new()
+            {
+                [BindingFlags.NonPublic | BindingFlags.Public] = 1,
+                [BindingFlags.Static] = 2,
+            };
+
+            string expected = @"{
+    ""public, non_public"": 1,
+    ""static"": 2
+}";
+
+            JsonTestHelper.AssertJsonEqual(expected, JsonSerializer.Serialize(dict, options));
```

Comment:
you might want to validate deserialization behavior here as well (throw or roundtrip)